data-import/guide-import-csv/guide-import-csv.adoc

```diff
@@ -10,282 +10,391 @@
+Other times, it is hard to find small syntax changes or make broad adjustments for files having a lot of data in them
 
-IMPORT DONE. Took: 58s 124ms
+For handling these types of situations or to speed up any type of data cleanup,there are a number of tools that help you check and validate your CSV data files.
+
+Basic tools, such as hexdump, vi, emacs, UltraEdit, and Notepad++ work well for handling shortcut-based commands for editing and manipulating files.
+However, these may not be the most efficient or user-friendly for those new to these tools.
+
+Thankfully, there are also other options available that can assist in data cleanup and formatting.
+We will list a few with links here.
+
+* link:https://csvkit.readthedocs.io/en/latest/[CSVKit^] - a set of Python tools that provides statistics (csvstat), search (csvgrep), and more for your CSV files.
+`csvstat` is very helpful to get an overview and statistics over your file, if you know your domain and data you can immediately spot inaccuracies.
+For instance, showing excess field length can indicate stray quotes.
+
+* link:http://csvlint.io/[CSVLint^] - an online service to validate CSV files.
+You can upload the file or provide an URL to load it.
+There is also the option to provide additional schema information.
+
+* link:https://www.papaparse.com/[Papa Parse^] - a comprehensive Javascript library for CSV parsing that allows you to stream CSV data and provides good, human-readable error reporting on issues.
+In the https://www.papaparse.com/demo[demo^] area, you can test your files or use the library directly.
+
+* link:/developer/desktop-csv-import/#inspect-files[Cypher] - what Cypher sees is what will be imported, so you can use that to your advantage.
+Using `LOAD CSV` without creating graph structure will just output samples, counts, or distributions to help you analyze exactly what is in the file and what Cypher will actually load.
+This format makes it possible to detect incorrect header column counts, delimiters, quotes, escapes, or header name spellings.
+
+[source, cypher]
+----
+// assert correct line count
+LOAD CSV FROM "file-url" AS line
+RETURN count(*);
+
+// check first few raw lines
+LOAD CSV FROM "file-url" AS line
+RETURN line
+LIMIT 5;
+
+// check first 5 line-sample with header-mapping
+LOAD CSV WITH HEADERS FROM "file-url" AS line
+RETURN line
+LIMIT 5;
 ----
 
 [#import-csv-resources]
-== Resources
-* link:{opsmanual}/tools/import/["Manual: Import Tool",role=docs]
-* link:{opsmanual}/tutorial/import-tool/["Manual: Import Tool Tutorial"]
-* link:/developer/data-modeling/guide-data-modeling[Intro to Graph Modeling]
-* http://www.intelliwareness.org/?p=583["Article: Importing Medicare Provider and Payment Data",role=blog]
-* http://blog.graphenedb.com/blog/2015/01/13/importing-data-into-neo4j-via-csv/[Article: LOAD CSV Intro with GrapheneDB]
+== CSV Import Resources
+
+* link:{opsmanual}/tools/import/[Manual: Import Tool^]
```

Comment:
did you check the links to the manual?

Reply:
Yes, I did! Everything looks good on those.